### PR TITLE
Enable custom index columns

### LINKED ISSUE #485: pyam.concat / pyam.append overwrites and not updates year parameter
the concat command has issues:

```
IAMC_IDX = ['model', 'scenario', 'region', 'variable', 'unit']

TEST_YEARS_PROJ = [2005, 2010]
TEST_YEARS_HIST = [2000]

proj = pd.DataFrame([
    ['model_a', 'scen_a', 'reg', 'PE', 'EJ/yr', 1, 6.],
],
    columns=IAMC_IDX + TEST_YEARS_PROJ,
)
proj = pyam.IamDataFrame(proj)

hist = pd.DataFrame([
    ['source_b', 'scen_b', 'reg', 'PE', 'EJ/yr', 0.5],
],
    columns=IAMC_IDX + TEST_YEARS_HIST,
)
hist = pyam.IamDataFrame(hist)

concat = pyam.concat([hist, proj])

print(concat)
print(concat.year)
```

gives me

```
 <class 'pyam.core.IamDataFrame'>
Index dimensions:
 * model    : model_a, source_b (2)
 * scenario : scen_a, scen_b (2)
Timeseries data coordinates:
   region   : reg (1)
   variable : PE (1)
   unit     : EJ/yr (1)
   year     : 2000, 2005, 2010 (3)
Meta indicators:
   exclude (bool) False (1)

 [2000]
```

(the latter should be [2000, 2005 and 2010]). The issue only arises for .year

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- ~Name of contributors Added to AUTHORS.rst~
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

Per discussions with related energy modelling projects (in particular OpenEnergyPlatform, SENTINEL), it became obvious that the restriction of the IAMC format to use `['model', 'scenario']` as the index (ie the organizing structure of timeseries data) is a limitation of the current package implementation.

Two specific examples:
 - The OpenEnergyPlatform is discussing a data model indexed on `['source', 'scenario']` to highlight that not all data are derived from models but can also come from *reference sources* <sup>[[citation needed]]()</sup>
 - When reading multiple versions of a model/scenario from an IIASA database, the API returns the index `['model', 'scenario', 'version]`, which can currently not be processed by pyam.

This PR extends the IamDataFrame class to accept an `index` argument such that timeseries data following the two examples above (or other index columns) can be handled by pyam. This PR is intended for discussion and as a first step to full-fledged support - only the initialization and `IamDataFrame.info()` are known to work at this point.

See this example below for an illustration

```python
import pandas as pd
import pyam

TEST_DF = pd.DataFrame([
    ['model_a', 'scen_a', 1, 'World', 'Primary Energy', 'EJ/yr', 1, 6.],
    ['model_a', 'scen_a', 2, 'World', 'Primary Energy|Coal', 'EJ/yr', 0.5, 3],
    ['model_a', 'scen_b', 3, 'World', 'Primary Energy', 'EJ/yr', 2, 7],
],
    columns=['source', 'scenario', 'version', 'region', 'variable', 'unit'] + [2005, 2010],
)

df = pyam.IamDataFrame(TEST_DF, index=['source', 'scenario', 'version'])
df.info()
```

This PR also solves an problem with the dynamically created attributes like `d.year` or `df.time`, closes #485

### CODE REVIEW

**Comment by danielhuppmann** [comment]
@ludee @christian-rli @tburandt @suvayu - I'd appreciate your comments!

File: pyam/core.py [hunk 1]
```diff
@@ -285,12 +287,19 @@ def index(self):
...
         """Return the list of (unique) scenario names"""
-        return get_index_levels(self.meta, 'scenario')
+        return self._get_meta_index_levels('scenario')
+
+    def _get_meta_index_levels(self, name):
+        """Return the list of a level from meta"""
+        if name in self.meta.index.names:
+            return get_index_levels(self.meta, name)
+        # in case of non-standard meta.index.names
+        raise AttributeError(f'Index `{name}` does not exist!')
```
Comment (suvayu) [inline pyam/core.py:302]:
```suggestion
        raise KeyError(f'Index `{name}` does not exist!')
```
or maybe `IndexError`

**Review by suvayu — COMMENTED** [review]
I left couple of minor suggestions, hope they are helpful.

File: pyam/core.py [hunk 2]
```diff
@@ -107,16 +109,18 @@ class IamDataFrame(object):
...
     """
-    def __init__(self, data, meta=None, **kwargs):
+    def __init__(self, data, meta=None, index=['model', 'scenario'], **kwargs):
         """Initialize an instance of an IamDataFrame"""
         if isinstance(data, IamDataFrame):
             if kwargs:
                 msg = 'Invalid arguments `{}` for initializing an IamDataFrame'
                 raise ValueError(msg.format(kwargs))
+            if index != data.index.names:
+                raise ValueError(f'Index {index} incompatible with:\n{data}')
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/core.py:119]: To clarify, `data` will show just a summary (via `print` & `info()`) and not the full timeseries data, see this test
https://github.com/IAMconsortium/pyam/blob/9151156ab9064776aa84ada01a0e0c2843601665/tests/test_core.py#L146

But you are right, shorter is even better!

File: pyam/core.py (hunk 1 above)
Comment (danielhuppmann) [reply pyam/core.py:302]: I really like the idea of using an `IndexError` when the index dimension doesn't exist, but I think `KeyError` is the more accurate error in this case (per https://docs.python.org/3/library/exceptions.html#IndexError)

**Comment by danielhuppmann** [comment]
> I left couple of minor suggestions, hope they are helpful.

Thank you @suvayu!

**Comment by christian-rli** [comment]
Frankly I am no expert on the OEP data model you are referring to @danielhuppmann . If you're talking about the data model developed in open_MODEX then @jh-RLI might be good at giving some technical insights. Other than that, the PR looks good to my non-expert eyes :)

**Comment by danielhuppmann** [comment]
per comment by @l-welder that this at least resolves her issue, I'll merge, and we'll support a custom index as a beta feature. Please start a new issue if the custom index doesn't work as expected.

**Pull request merged**